# 💻 **Introduction**

Sebelum sebuah data dapat dianalisis atau digunakan untuk mengambil keputusan kebijakan tentu data tersebut perlu untuk diekstraksi terlebih dahulu. <b>Ekstraksi data</b> adalah sebuah proses pengumpulan atau pengambilan jenis data yang berbeda dari berbagai sumber, banyak di antaranya mungkin tidak terorganisir dengan baik atau tidak terstruktur sama sekali. Proses ini memungkinkan untuk mengkonsolidasikan, memproses, dan menyempurnakan data sehingga nantinya dapat disimpan di lokasi terpusat untuk diubah atau diproses lebih lanjut.

Python sebagai salah satu <i>tools</i> yang <i>powerful</i> untuk pengolahan data menyediakan berbagai <i>library</i> untuk mendukung proses ekstraksi data. Salah satu library paling banyak digunakan adalah <b>pandas</b>.

Pandas memungkinkan untuk membaca data dari berbagai jenis ekstensi seperti data yang bersumber dari database (sql), <i>excel</i> (<b>.xls</b> atau <b>.xlsx</b>), <i>separated value file</i> (<b>.csv</b>, <b>.tsv</b>, dsb) lalu merubahnya menjadi satu format struktur data yang sama yang disebut <b>DataFrame</b>. DataFrame inilah yang nantinya dapat kita lakukan proses <b>Data Manipulation</b><br><br>

<p align="center">
<img src="https://drive.google.com/uc?export=view&id=18NW3gMMTJjkwwpZX5TXi9DTteXxGAoxa" width="600" height="300">
<p align="center"><b>Gambar 1. Proses Ekstraksi Data dengan <i>pandas</i></b></p>

<br><br>
<p align="right"><sup>docs : <i><a href="https://shorturl.at/xyZ69">https://pandas.pydata.org/docs/user_guide/io.html</i></sup></a></p>


## 🏃🏻‍♂️ **1. First Step : Pandas**

Untuk dapat menggunakan library pandas pada python, perlu untuk mengimpor library tersebut dengan cara mengetikan sintaks berikut

<p align="center"><code><b>import pandas as pd</b></code></p>

lalu jalankan (<i>run</i>) program.

<br><br>
<p align="left"><sup>docs : <i><a href="https://shorturl.at/kzG06">https://pandas.pydata.org/docs/getting_started/install.html</i></sup></a></p>

### 👨‍💻 Task 1

Ketikan kode `import pandas as pd` pada prompt dibawah ini lalu jalankan

In [23]:
# Write your code below here & exec!
import pandas as pd


# 🥨 **Struktur Data Pandas**

* Terdapat dua struktur data utama pada library pandas yang disebut dengan Series dan DataFrame.
<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="289">
<p align="center"><b>Gambar 2. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>

<p align="right"><sup>docs : <i><a href="https://shorturl.at/dezQY">http://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html</i></sup></a></p>

##🛺 **2. Series**

* <a href="">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.

* Untuk membuat Series gunakan sintaks berikut

<p align="center"><code><b>pd.Series(data)</b></code></p>

<ul>
  data dapat berupa <i>array-like</i>, <i>Iterable</i>, <i>dict</i>, atau <i>scalar</i>
</ul>

<p align="right"><sup>docs : <i><a href="https://shorturl.at/gmtG2">https://pandas.pydata.org/docs/reference/api/pandas.Series.html</i></sup></a></p>



### 👨‍💻 Task 2 - A

<i>Age of Accounts Receivable</i> (Aging AR) adalah istilah yang merujuk pada lamanya piutang dagang yang masih belum tertagih pada suatu periode tertentu. DQTech mecatat distribusi Aging AR tersaji dari list data sebagai berikut \[78, 91, 40, 50, 95\].

Dari data tersebut buatlah sebuah Series yang disimpan pada variabel dengan nama `ar_aging` lalu tampilkan outputnya!

In [24]:
# Inisialisasi data
data_age_of_AR = [78, 91, 40, 50, 95]

# Write your code below here & exec!
ar_aging = pd.Series(data_age_of_AR)

# Menampilkan hasil
print(ar_aging)

0    78
1    91
2    40
3    50
4    95
dtype: int64


### 👨‍💻 Task 2 - B

Dari data Aging AR yang tercatat, akan diberi detail data sebagai berikut
* Cust 1 memiliki Aging AR sebesar 78
* Cust 2 memiliki Aging AR sebesar 91
* Cust 3 memiliki Aging AR sebesar 40
* Cust 4 memiliki Aging AR sebesar 50
* Cust 5 memiliki Aging AR sebesar 95

Dari data tersebut buatlah sebuah Series yang disimpan pada variabel dengan nama `detail_ar_aging` lalu tampilkan outputnya!

In [25]:
# Inisialisasi data
data_age_of_AR = {
    'Cust1' : 78,
    'Cust2' : 91,
    'Cust3' : 40,
    'Cust4' : 50,
    'Cust5' : 95
}

# Write your code below here & exec!
detail_ar_aging = pd.Series(data_age_of_AR)

# Tampilkan hasilnya
display(detail_ar_aging)

Cust1    78
Cust2    91
Cust3    40
Cust4    50
Cust5    95
dtype: int64

In [26]:
type(detail_ar_aging)

pandas.core.series.Series

## 🚴🏻‍♂️ **3. DataFrame**

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame">DataFrame</a> adalah struktur data 2 dimensi yang terdiri dari baris dan kolom (seperti tabel) dan dapat menyimpan data dari berbagai jenis (termasuk karakter, bilangan bulat, bilangan real, data kategorikal, dan lainnya).

<p align="center">
<img src="https://www.w3resource.com/w3r_images/pandas-data-frame.svg" width="600" height="300">
<p align="center"><b>Gambar 2. DataFrame Pandas</b></p>
<br>

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.


<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="300">
<p align="center"><b>Gambar 3. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>




### 👨‍💻 Task 3

Data karyawan DQTech disajikan pada dictionary berikut

```
details = {
    'Name' : ['Aksara', 'Antara', 'Sendja', 'Sunyi'],
    'Age' : [23, 21, 22, 21],
    'Role' : ['Jr. Data Scientist', 'Data Engineer', 'Data Analyst', 'Data Scientist']
}
```

konversi dictionary tersebut menjadi DataFrame pandas

In [27]:
# Import library
import pandas as pd

# Initialize Data
details = {
    'Name' : ['Aksara', 'Antara', 'Sendja', 'Sunyi'],
    'Age' : [23, 21, 22, 21],
    'Role' : ['Jr. Data Scientist', 'Data Engineer', 'Data Analyst', 'Data Scientist']
}

# Write your code below here & exec!
dqtech_employee = pd.DataFrame(details)

# Tampilkan data
display(dqtech_employee)

Name  Age                Role
0  Aksara   23  Jr. Data Scientist
1  Antara   21       Data Engineer
2  Sendja   22        Data Analyst
3   Sunyi   21      Data Scientist

## ♖ **4. Parts Of DataFrame**

Untuk lebih memahami objek DataFrame, perlu diketahui bahwa DataFrame terdiri dari tiga komponen sebagai atribut berikut :

<ul>1️⃣ .values : Nilai - nilai yang tertera pada DataFrame</ul>
<ul>2️⃣ .columns : Nama kolom yang ada pada DataFrame</ul>
<ul>3️⃣ .index : Nama atau nomor index pada DataFrame</ul>



### 👨‍💻 Task 4

Pada DataFrame yang telah dibuat pada Task 3, coba buatkan list nama kolomnya!

In [28]:
# Write your code below here & exec!
kolom = dqtech_employee.columns

# Tampilkan hasilnya
print(kolom)


Index(['Name', 'Age', 'Role'], dtype='object')


---


# 📊 **Data Manipulation**

Data Manipulation adalah sebuah metode pengorganisasian data agar lebih mudah dibaca, terstruktur dan agar lebih mudah diolah untuk kebutuhan lain kedepannya. Beberapa hal yang biasa dilakukan saat data manipulation adalah sebagai berikut :


<ul>1️⃣ Column selections</ul>
<ul>2️⃣ Filter the data</ul>
<ul>3️⃣ Create a new column or change it</ul>
<ul>4️⃣ Drop a column</ul>
<ul>5️⃣ Apply custom functions</ul>
<ul>6️⃣ Join / Merge the data</ul>
<ul>7️⃣ etc</ul>


## ⚓️ **5. Data Extraction**

* Dalam konteks data extraction, Pandas dapat digunakan untuk membaca data dari berbagai sumber, seperti file CSV, Excel, SQL database, dan lainnya, dan mengubahnya menjadi DataFrames untuk analisis lebih lanjut.

* Beberapa fungsi dan metode Pandas yang umum digunakan untuk data extraction adalah sebagai berikut:

<ul>
<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html">pd.read_csv()</a></code>: Digunakan untuk membaca data dari file CSV dan mengonversinya menjadi DataFrame.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html">pd.read_excel()</a></code>: Membaca data dari file Excel dan membuat DataFrame dari lembar kerja atau tabel tertentu.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html">pd.read_sql()</a></code>: Mengambil data dari basis data SQL menggunakan SQL query dan menyimpannya dalam bentuk DataFrame.
</ul>


### 👨‍💻 Task 5

Untuk melakukan proses ekstraksi data, lakukan instruksi berikut :     

1.   Persiapkan dataset SuperStore.xlsx
2.   Upload ke google colab dengan cara pilih File pada bagian kiri laman (icon 📁), kemudian klik kanan dan pilih upload
3.   Lakukan ekstraksi data excel (.xlsx) dengan library pandas, assign hasilnya ke sebuah variabel bernama <code>data_superstore</code>
4.   Tampilkan hasilnya






In [29]:
# Import library
import pandas as pd

# Proses ekstraksi data
data_superstore = pd.read_excel('/content/Superstore.xlsx')

# Tampilkeun hasilnya
display(data_superstore)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
0          1  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
1          2  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
2          3  CA-2016-138688 2016-06-12 2016-06-16    Second Class   
3          4  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
4          5  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
...      ...             ...        ...        ...             ...   
9989    9990  CA-2014-110422 2014-01-21 2014-01-23    Second Class   
9990    9991  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9991    9992  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9992    9993  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9993    9994  CA-2017-119914 2017-05-04 2017-05-09    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United States        Henderson   
1       CG-12520       Claire Gute   Consumer  United States        Henderson   
2       DV-13045   Darrin Van Huff  Corporate  United States      Los Angeles   
3       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
4       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...          ...               ...        ...            ...              ...   
9989    TB-21400  Tom Boeckenhauer   Consumer  United States            Miami   
9990    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9991    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9992    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9993    CC-12220      Chris Cortes   Consumer  United States      Westminster   

      ... Postal Code  Region       Product ID         Category Sub-Category  \
0     ...       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1     ...       42420   South  FUR-CH-10000454        Furniture       Chairs   
2     ...       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3     ...       33311   South  FUR-TA-10000577        Furniture       Tables   
4     ...       33311   South  OFF-ST-10000760  Office Supplies      Storage   
...   ...         ...     ...              ...              ...          ...   
9989  ...       33180   South  FUR-FU-10001889        Furniture  Furnishings   
9990  ...       92627    West  FUR-FU-10000747        Furniture  Furnishings   
9991  ...       92627    West  TEC-PH-10003645       Technology       Phones   
9992  ...       92627    West  OFF-PA-10004041  Office Supplies        Paper   
9993  ...       92683    West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name     Sales  Quantity  \
0                     Bush Somerset Collection Bookcase  261.9600         2   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2     Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3         Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                        Eldon Fold 'N Roll Cart System   22.3680         2   
...                                                 ...       ...       ...   
9989                             Ultra Door Pull Handle   25.2480         3   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...   91.9600         2   
9991                              Aastra 57i VoIP phone  258.5760         2   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.6000         4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.

## ⛑ **6. Inspecting Data Information**

* Untuk dapat menginspeksi informasi umum pada data dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.info()</b></code></p>

<ul>Ketika dijalankan, sintaks tersebut akan menghasilkan output informasi umum pada data seperti banyak kolom, nama kolom beserta tipe data dan jumlah data tak kosong (<i>non-null value</i>)</ul><br>

<p align="right"><sup>docs : <i><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html</i></sup></a></p>


### 👨‍💻 Task 6

Inspeksi informasi umum pada data yang telah diekstraksi dan telah disimpan pada variabel <code>data_superstore</code>. Setelah itu coba uraikan informasi apa yang bisa anda dapatkan setelah melakukan inspeksi informasi umum pada <code>data_superstore</code>?

In [30]:
# Write your code below here & exec!
data_superstore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

In [31]:
dt_superstore = data_superstore.drop(columns = ['Row ID'])

In [32]:
dt_superstore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Order ID       9994 non-null   object        
 1   Order Date     9994 non-null   datetime64[ns]
 2   Ship Date      9994 non-null   datetime64[ns]
 3   Ship Mode      9994 non-null   object        
 4   Customer ID    9994 non-null   object        
 5   Customer Name  9994 non-null   object        
 6   Segment        9994 non-null   object        
 7   Country        9994 non-null   object        
 8   City           9994 non-null   object        
 9   State          9994 non-null   object        
 10  Postal Code    9994 non-null   int64         
 11  Region         9994 non-null   object        
 12  Product ID     9994 non-null   object        
 13  Category       9994 non-null   object        
 14  Sub-Category   9994 non-null   object        
 15  Product Name   9994 n

In [33]:
dt_superstore.head(10)

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
5  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
6  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
7  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
8  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
9  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
5  Brosina Hoffman   Consumer  United States      Los Angeles  California   
6  Brosina Hoffman   Consumer  United States      Los Angeles  California   
7  Brosina Hoffman   Consumer  United States      Los Angeles  California   
8  Brosina Hoffman   Consumer  United States      Los Angeles  California   
9  Brosina Hoffman   Consumer  United States      Los Angeles  California   

   Postal Code Region       Product ID         Category Sub-Category  \
0        42420  South  FUR-BO-10001798        Furniture    Bookcases   
1        42420  South  FUR-CH-10000454        Furniture       Chairs   
2        90036   West  OFF-LA-10000240  Office Supplies       Labels   
3        33311  South  FUR-TA-10000577        Furniture       Tables   
4        33311  South  OFF-ST-10000760  Office Supplies      Storage   
5        90032   West  FUR-FU-10001487        Furniture  Furnishings   
6        90032   West  OFF-AR-10002833  Office Supplies          Art   
7        90032   West  TEC-PH-10002275       Technology       Phones   
8        90032   West  OFF-BI-10003910  Office Supplies      Binders   
9        90032   West  OFF-AP-10002892  Office Supplies   Appliances   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   
5  Eldon Expressions Wood and Plastic Desk Access...   48.8600         7   
6                                         Newell 322    7.2800         4   
7                     Mitel 5320 IP Phone VoIP phone  907.1520         6   
8  DXL Angle-View Binders with Locking Rings by S...   18.5040         3   
9                   Belkin F5C206VTEL 6 Outlet Surge  114.9000         5   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  
6      0.00    1.9656  
7      0.20   90.7152  
8      0.20    5.7825  
9      0.00   34.4700

## 🔱 **7. Data Selection**

* Untuk memilih kolom tertentu pada sebuah DataFrame dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.loc[baris : nama_kolom]</b></code></p>

* <code>baris</code>
<ul>
  <li>Gunakan simbol ( : ) untuk menampilkan semua data</li>
  <li>Gunakan simbol (m : n) untuk menampilkan data dari index ke - m sampai dengan index ke - n</li>
</ul>

* <code>nama_kolom</code>
<ul>
  <li>Gunakan string untuk mengakses satu kolom saja (<i>return : Series</i>)</li>
  <li>Gunakan list untuk mengakses lebih dari satu kolom  (<i>return : DataFrame</i>)</li>
</ul>


### 👨‍💻 Task 7 - A

Pada DataFrame <code>data_superstore</code>, simpan 10 data teratas ke dalam variabel **order_id** untuk kolom **Order_ID** saja. Kemudian tampilkan data tersebut!

In [34]:
# Write your code below here & exec!
order_id = dt_superstore.loc[:10, 'Order ID']

display(order_id)

0     CA-2016-152156
1     CA-2016-152156
2     CA-2016-138688
3     US-2015-108966
4     US-2015-108966
5     CA-2014-115812
6     CA-2014-115812
7     CA-2014-115812
8     CA-2014-115812
9     CA-2014-115812
10    CA-2014-115812
Name: Order ID, dtype: object

### 👨‍💻 Task 7 - B

Pada DataFrame <code>data_superstore</code>, tampilkan data dengan index ke-2 sampai dengan index ke-5 kolom **Order_ID**, **Sales**, **Profit** saja

In [35]:
# Write your code below here & exec!
col_selection = ['Order ID', 'Sales', 'Profit']

data_sales = dt_superstore.loc[2:5, ['Order ID', 'Sales', 'Profit']]

display(data_sales)

Order ID     Sales    Profit
2  CA-2016-138688   14.6200    6.8714
3  US-2015-108966  957.5775 -383.0310
4  US-2015-108966   22.3680    2.5164
5  CA-2014-115812   48.8600   14.1694

atau secara sederhana anda dapat menggunakan `DataFrame[nama_kolom]` untuk menghasilkan sebuah Series satu kolom dan `DataFrame[[nama_kolom1, nama_kolom2, ...]]` untuk menseleksi banyak kolom kedalam bentuk DataFrame

In [36]:
col_selection = ['Order ID', 'Sales', 'Profit']

data_sales = dt_superstore[['Order ID', 'Sales', 'Profit']].head(6)

display(data_sales)

Order ID     Sales    Profit
0  CA-2016-152156  261.9600   41.9136
1  CA-2016-152156  731.9400  219.5820
2  CA-2016-138688   14.6200    6.8714
3  US-2015-108966  957.5775 -383.0310
4  US-2015-108966   22.3680    2.5164
5  CA-2014-115812   48.8600   14.1694

### 👨‍💻 Task 7 - C

Pada DataFrame <code>data_superstore</code>, simpan ke dalam variabel **order_id** untuk kolom **Order_ID** saja (gunakan seleksi sederhana). Kemudian tampilkan data tersebut!

In [37]:
# Write your code below here & exec!
order_id = dt_superstore['Order ID']

display(order_id)

0       CA-2016-152156
1       CA-2016-152156
2       CA-2016-138688
3       US-2015-108966
4       US-2015-108966
             ...      
9989    CA-2014-110422
9990    CA-2017-121258
9991    CA-2017-121258
9992    CA-2017-121258
9993    CA-2017-119914
Name: Order ID, Length: 9994, dtype: object

## 🎵 **8. Conditions**

* Salah satu teknik paling sederhana untuk menemukan bagian mana dari kumpulan data yang menarik adalah menemukan subset baris yang cocok dengan beberapa kriteria untuk difilter atau dipilih. Untuk melakukan hal ini cukup mudah yakni dengan menuliskan kondisi seperti berikut

<p align="center"><code><b>DataFrame.loc[conditions]</b></code></p>

* Conditions dapat dibentuk dari 1 atau lebih kriteria boolean dan dihubungkan dengan operator logika and (&), or (|) atau not (~)

<br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html</i>

### 👨‍💻 Task 8 - A

Pada DataFrame `data_superstore` tampilkan semua record data yang terjadi di 'Los Angeles' saja

In [38]:
# Assign conditions
conditions = dt_superstore['City'] == 'Los Angeles'

# Filter
data_LA = dt_superstore.loc[conditions]

# Tampilkan hasil
display(data_LA)

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
2     CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
5     CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
6     CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
7     CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
8     CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
...              ...        ...        ...             ...         ...   
9976  US-2016-103674 2016-12-06 2016-12-10  Standard Class    AP-10720   
9977  US-2016-103674 2016-12-06 2016-12-10  Standard Class    AP-10720   
9978  US-2016-103674 2016-12-06 2016-12-10  Standard Class    AP-10720   
9979  US-2016-103674 2016-12-06 2016-12-10  Standard Class    AP-10720   
9986  CA-2016-125794 2016-09-29 2016-10-03  Standard Class    ML-17410   

        Customer Name      Segment        Country         City       State  \
2     Darrin Van Huff    Corporate  United States  Los Angeles  California   
5     Brosina Hoffman     Consumer  United States  Los Angeles  California   
6     Brosina Hoffman     Consumer  United States  Los Angeles  California   
7     Brosina Hoffman     Consumer  United States  Los Angeles  California   
8     Brosina Hoffman     Consumer  United States  Los Angeles  California   
...               ...          ...            ...          ...         ...   
9976       Anne Pryor  Home Office  United States  Los Angeles  California   
9977       Anne Pryor  Home Office  United States  Los Angeles  California   
9978       Anne Pryor  Home Office  United States  Los Angeles  California   
9979       Anne Pryor  Home Office  United States  Los Angeles  California   
9986     Maris LaWare     Consumer  United States  Los Angeles  California   

      Postal Code Region       Product ID         Category Sub-Category  \
2           90036   West  OFF-LA-10000240  Office Supplies       Labels   
5           90032   West  FUR-FU-10001487        Furniture  Furnishings   
6           90032   West  OFF-AR-10002833  Office Supplies          Art   
7           90032   West  TEC-PH-10002275       Technology       Phones   
8           90032   West  OFF-BI-10003910  Office Supplies      Binders   
...           ...    ...              ...              ...          ...   
9976        90032   West  TEC-PH-10002496       Technology       Phones   
9977        90032   West  OFF-FA-10003467  Office Supplies    Fasteners   
9978        90032   West  OFF-BI-10003727  Office Supplies      Binders   
9979        90032   West  OFF-BI-10002026  Office Supplies      Binders   
9986        90008   West  TEC-AC-10003399       Technology  Accessories   

                                           Product Name    Sales  Quantity  \
2     Self-Adhesive Address Labels for Typewriters b...   14.620         2   
5     Eldon Expressions Wood and Plastic Desk Access...   48.860         7   
6                                            Newell 322    7.280         4   
7                        Mitel 5320 IP Phone VoIP phone  907.152         6   
8     DXL Angle-View Binders with Locking Rings by S...   18.504         3   
...                                                 ...      ...       ...   
9976                                       Cisco SPA301  249.584         2   
9977           Alliance Big Bands Rubber Bands, 12/Pack   13.860         7   
9978  Avery Durable Slant Ring Binders With Label Ho...   13.376         4   
9979                  Ibico Recycled Linen-Style Covers  437.472        14   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...   36.240         1   

      Discount    Profit  
2          0.0    6.8714  
5          0.0   14.1694  
6          0.0    1.9656  
7          0.2   90.7152  
8          0.2    5.7825  
...        ...       ...  
9976       0.2   31.1980  
9977       0.0    0.0000  
9978       0.2    4.6816  
9979       0.2  153.1152  
9986       0.0   15.2208  

[747 rows x 20 columns]

### 👨‍💻 Task 8 - B

Pada DataFrame `data_superstore` tampilkan semua record data yang terjadi di 'Los Angeles' untuk kategori Technology

In [39]:
# Write your code below here & exec!

# Assign conditions
conditions1 = dt_superstore['City'] == 'Los Angeles'
conditions2 = dt_superstore['Category'] == 'Technology'

# Filter
data_LA_tech = dt_superstore.loc[conditions1 & conditions2]

# Tampilkeun
display(data_LA_tech)

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
7     CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
11    CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
26    CA-2016-121755 2016-01-16 2016-01-20    Second Class    EH-13945   
62    CA-2015-135545 2015-11-24 2015-11-30  Standard Class    KM-16720   
90    CA-2016-109806 2016-09-17 2016-09-22  Standard Class    JS-15685   
...              ...        ...        ...             ...         ...   
9884  CA-2014-112291 2014-04-03 2014-04-08  Standard Class    KE-16420   
9885  CA-2014-112291 2014-04-03 2014-04-08  Standard Class    KE-16420   
9973  US-2016-103674 2016-12-06 2016-12-10  Standard Class    AP-10720   
9976  US-2016-103674 2016-12-06 2016-12-10  Standard Class    AP-10720   
9986  CA-2016-125794 2016-09-29 2016-10-03  Standard Class    ML-17410   

        Customer Name      Segment        Country         City       State  \
7     Brosina Hoffman     Consumer  United States  Los Angeles  California   
11    Brosina Hoffman     Consumer  United States  Los Angeles  California   
26      Eric Hoffmann     Consumer  United States  Los Angeles  California   
62       Kunst Miller     Consumer  United States  Los Angeles  California   
90           Jim Sink    Corporate  United States  Los Angeles  California   
...               ...          ...            ...          ...         ...   
9884  Katrina Edelman    Corporate  United States  Los Angeles  California   
9885  Katrina Edelman    Corporate  United States  Los Angeles  California   
9973       Anne Pryor  Home Office  United States  Los Angeles  California   
9976       Anne Pryor  Home Office  United States  Los Angeles  California   
9986     Maris LaWare     Consumer  United States  Los Angeles  California   

      Postal Code Region       Product ID    Category Sub-Category  \
7           90032   West  TEC-PH-10002275  Technology       Phones   
11          90032   West  TEC-PH-10002033  Technology       Phones   
26          90049   West  TEC-AC-10003027  Technology  Accessories   
62          90004   West  TEC-AC-10004633  Technology  Accessories   
90          90036   West  TEC-PH-10004093  Technology       Phones   
...           ...    ...              ...         ...          ...   
9884        90008   West  TEC-AC-10001284  Technology  Accessories   
9885        90008   West  TEC-AC-10000736  Technology  Accessories   
9973        90032   West  TEC-PH-10004080  Technology       Phones   
9976        90032   West  TEC-PH-10002496  Technology       Phones   
9986        90008   West  TEC-AC-10003399  Technology  Accessories   

                                           Product Name    Sales  Quantity  \
7                        Mitel 5320 IP Phone VoIP phone  907.152         6   
11        Konftel 250 Conference phone - Charcoal black  911.424         4   
26     Imation 8GB Mini TravelDrive USB 2.0 Flash Drive   90.570         3   
62    Verbatim 25 GB 6x Blu-ray Single Layer Recorda...   13.980         2   
90                                   Panasonic Kx-TS550   73.584         2   
...                                                 ...      ...       ...   
9884  Enermax Briskie RF Wireless Keyboard and Mouse...   62.310         3   
9885                     Logitech G600 MMO Gaming Mouse  159.980         2   
9973                           Avaya 5410 Digital phone  271.960         5   
9976                                       Cisco SPA301  249.584         2   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...   36.240         1   

      Discount   Profit  
7          0.2  90.7152  
11         0.2  68.3568  
26         0.0  11.7741  
62         0.0   6.1512  
90         0.2   8.2782  
...        ...      ...  
9884       0.0  22.4316  
9885       0.0  57.5928  
9973       0.2  27.1960  
9976       0.2  31.1980  
9986       0.0  15.2208  

[150 rows x 20 columns]

### 👨‍💻 Task 8 - C

Pada DataFrame `data_superstore` tampilkan record data yang terjadi di 'Los Angeles' untuk kategori Technology. Namun hanya tampilkan kolom 'Order_ID', 'City', 'Category' dan 'Profit'

In [40]:
# Assign conditions
conditions1 = dt_superstore['City'] == 'Los Angeles'
conditions2 = dt_superstore['Category'] == 'Technology'
list_shown = ['Order ID', 'City', 'Category', 'Profit']

# Filter
data_LA_tech = dt_superstore.loc[conditions1 & conditions2, list_shown]

# Tampilkeun
display(data_LA_tech)

Order ID         City    Category   Profit
7     CA-2014-115812  Los Angeles  Technology  90.7152
11    CA-2014-115812  Los Angeles  Technology  68.3568
26    CA-2016-121755  Los Angeles  Technology  11.7741
62    CA-2015-135545  Los Angeles  Technology   6.1512
90    CA-2016-109806  Los Angeles  Technology   8.2782
...              ...          ...         ...      ...
9884  CA-2014-112291  Los Angeles  Technology  22.4316
9885  CA-2014-112291  Los Angeles  Technology  57.5928
9973  US-2016-103674  Los Angeles  Technology  27.1960
9976  US-2016-103674  Los Angeles  Technology  31.1980
9986  CA-2016-125794  Los Angeles  Technology  15.2208

[150 rows x 4 columns]

## 📎 **9. Sorting Row**

Menemukan data yang menarik dalam sebuah DataFrame seringkali lebih mudah jika dilakukan pengurutan baris datanya. Pengurutan dibagi menjadi dua yakni pengurutan secara `Ascending` (pengurutan naik A → Z) dan juga pengurutan secara `Descending` (pengurutan turun Z → A). Eksekusi sintaks berikut :

<p align="center"><code><b>DataFrame.sort_values(by, ascending=True, ignore_index=False)</b></code></p>

Jika ingin mengurutkan DataFrame pandas dengan memperhatikan argumen

<ul>
<li><b><i>by</i></b> sebagai satu atau lebih kolom yang ingin dijadikan patokan pengurutan</li>

<li><b><i>ascending</i></b> dapat diisikan True jika ingin terurut naik atau False jika ingin terurut turun (dafault : True)</li>

<li><b><i>ignore_index</i></b> dapat diisikan True jika ingin index masih terurut atau False jika indexnya masih dipertahankan sesuai data awal (dafault : False)</li>

</ul><br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html</i>


### 👨‍💻 Task 9 - A

Pada DataFrame `data_superstore`, tampilkan semua record data namun urutkan secara ascending pada kolom Order_ID

In [41]:
# Write your code below here & exec!
dt_superstore.sort_values(by = 'Order ID', ascending=True, ignore_index=False)

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
2717  CA-2014-100006 2014-09-07 2014-09-13  Standard Class    DK-13375   
6288  CA-2014-100090 2014-07-08 2014-07-12  Standard Class    EB-13705   
6287  CA-2014-100090 2014-07-08 2014-07-12  Standard Class    EB-13705   
9514  CA-2014-100293 2014-03-14 2014-03-18  Standard Class    NF-18475   
3083  CA-2014-100328 2014-01-28 2014-02-03  Standard Class    JC-15340   
...              ...        ...        ...             ...         ...   
5932  US-2017-169551 2017-07-07 2017-07-09     First Class    RL-19615   
5933  US-2017-169551 2017-07-07 2017-07-09     First Class    RL-19615   
5934  US-2017-169551 2017-07-07 2017-07-09     First Class    RL-19615   
5929  US-2017-169551 2017-07-07 2017-07-09     First Class    RL-19615   
5931  US-2017-169551 2017-07-07 2017-07-09     First Class    RL-19615   

         Customer Name      Segment        Country           City  \
2717       Dennis Kane     Consumer  United States  New York City   
6288        Ed Braxton    Corporate  United States  San Francisco   
6287        Ed Braxton    Corporate  United States  San Francisco   
9514  Neil Französisch  Home Office  United States   Jacksonville   
3083   Jasper Cacioppo     Consumer  United States  New York City   
...                ...          ...            ...            ...   
5932         Rob Lucas     Consumer  United States   Philadelphia   
5933         Rob Lucas     Consumer  United States   Philadelphia   
5934         Rob Lucas     Consumer  United States   Philadelphia   
5929         Rob Lucas     Consumer  United States   Philadelphia   
5931         Rob Lucas     Consumer  United States   Philadelphia   

             State  Postal Code Region       Product ID         Category  \
2717      New York        10024   East  TEC-PH-10002075       Technology   
6288    California        94122   West  OFF-BI-10001597  Office Supplies   
6287    California        94122   West  FUR-TA-10003715        Furniture   
9514       Florida        32216  South  OFF-PA-10000176  Office Supplies   
3083      New York        10024   East  OFF-BI-10000343  Office Supplies   
...            ...          ...    ...              ...              ...   
5932  Pennsylvania        19120   East  OFF-ST-10004835  Office Supplies   
5933  Pennsylvania        19120   East  TEC-AC-10002018       Technology   
5934  Pennsylvania        19120   East  TEC-AC-10003033       Technology   
5929  Pennsylvania        19120   East  FUR-BO-10001519        Furniture   
5931  Pennsylvania        19120   East  TEC-PH-10001363       Technology   

     Sub-Category                                       Product Name    Sales  \
2717       Phones                                  AT&T EL51110 DECT  377.970   
6288      Binders  Wilson Jones Ledger-Size, Piano-Hinge Binder, ...  196.704   
6287       Tables            Hon 2111 Invitation Series Corner Table  502.488   
9514        Paper                                         Xerox 1887   91.056   
3083      Binders  Pressboard Covers with Storage Hooks, 9 1/2" x...    3.928   
...           ...                                                ...      ...   
5932      Storage                  Plastic Stacking Crates & Casters   13.392   
5933  Accessories              AmazonBasics 3-Button USB Wired Mouse   16.776   
5934  Accessories  Plantronics CS510 - Over-the-Head monaural Wir...  527.920   
5929    Bookcases            O'Sullivan 3-Shelf Heavy-Duty Bookcases   87.210   
5931       Phones                                    Apple iPhone 5S  683.988   

      Quantity  Discount    Profit  
2717         3       0.0  109.6113  
6288         6       0.2   68.8464  
6287         3       0.2  -87.9354  
9514         6       0.2   31.8696  
3083         1       0.2    1.3257  
...        ...       ...       ...  
5932         3       0.2    1.0044  
5933         3       0.2    4.8231  
5934         2       0.2   85.7870  
5929         3       0.5  -45.3492  
5

### 👨‍💻 Task 9 - B

Pada DataFrame `data_superstore`, tampilkan semua record data namun urutkan secara descending pada kolom Profit dan urutkan secara ascending pada kolom Quantity

In [42]:
# Write your code below here & exec!
sorted_data = dt_superstore.sort_values(by = ['Profit', 'Quantity'], ascending=[False, True], ignore_index=False)

display(sorted_data)

Order ID Order Date  Ship Date       Ship Mode Customer ID  \
6826  CA-2016-118689 2016-10-02 2016-10-09  Standard Class    TC-20980   
8153  CA-2017-140151 2017-03-23 2017-03-25     First Class    RB-19360   
4190  CA-2017-166709 2017-11-17 2017-11-22  Standard Class    HL-15040   
9039  CA-2016-117121 2016-12-17 2016-12-21  Standard Class    AB-10105   
4098  CA-2014-116904 2014-09-23 2014-09-28  Standard Class    SC-20095   
...              ...        ...        ...             ...         ...   
4991  US-2017-122714 2017-12-07 2017-12-13  Standard Class    HG-14965   
3011  CA-2017-134845 2017-04-17 2017-04-23  Standard Class    SR-20425   
9774  CA-2014-169019 2014-07-26 2014-07-30  Standard Class    LF-17185   
683   US-2017-168116 2017-11-04 2017-11-04        Same Day    GT-14635   
7772  CA-2016-108196 2016-11-25 2016-12-02  Standard Class    CS-12505   

       Customer Name      Segment        Country         City           State  \
6826    Tamara Chand    Corporate  United States    Lafayette         Indiana   
8153    Raymond Buch     Consumer  United States      Seattle      Washington   
4190    Hunter Lopez     Consumer  United States       Newark        Delaware   
9039   Adrian Barton     Consumer  United States      Detroit        Michigan   
4098    Sanjit Chand     Consumer  United States  Minneapolis       Minnesota   
...              ...          ...            ...          ...             ...   
4991   Henry Goldwyn    Corporate  United States      Chicago        Illinois   
3011  Sharelle Roach  Home Office  United States   Louisville        Colorado   
9774     Luke Foster     Consumer  United States  San Antonio           Texas   
683   Grant Thornton    Corporate  United States   Burlington  North Carolina   
7772   Cindy Stewart     Consumer  United States    Lancaster            Ohio   

      Postal Code   Region       Product ID         Category Sub-Category  \
6826        47905  Central  TEC-CO-10004722       Technology      Copiers   
8153        98115     West  TEC-CO-10004722       Technology      Copiers   
4190        19711     East  TEC-CO-10004722       Technology      Copiers   
9039        48205  Central  OFF-BI-10000545  Office Supplies      Binders   
4098        55407  Central  OFF-BI-10001120  Office Supplies      Binders   
...           ...      ...              ...              ...          ...   
4991        60653  Central  OFF-BI-10001120  Office Supplies      Binders   
3011        80027     West  TEC-MA-10000822       Technology     Machines   
9774        78207  Central  OFF-BI-10004995  Office Supplies      Binders   
683         27217    South  TEC-MA-10004125       Technology     Machines   
7772        43130     East  TEC-MA-10000418       Technology     Machines   

                                          Product Name      Sales  Quantity  \
6826             Canon imageCLASS 2200 Advanced Copier  17499.950         5   
8153             Canon imageCLASS 2200 Advanced Copier  13999.960         4   
4190             Canon imageCLASS 2200 Advanced Copier  10499.970         3   
9039  GBC Ibimaster 500 Manual ProClick Binding System   9892.740        13   
4098              Ibico EPK-21 Electric Binding System   9449.950         5   
...                                                ...        ...       ...   
4991              Ibico EPK-21 Electric Binding System   1889.990         5   
3011         Lexmark MX611dhe Monochrome Laser Printer   2549.985         5   
9774         GBC DocuBind P400 Electric Binding System   2177.584         8   
683          Cubify CubeX 3D Printer Triple Head Print   7999.980         4   
7772         Cubify CubeX 3D Printer Double Head Print   4499.985         5   

      Discount     Profit  
6826       0.0  8399.9760  
8153       0.0  6719.9808  
4190       0.0  5039.9856  
9039       0.0  4946.3700  
4098       0.0  4630.4755  
...        ...        ...  
4991       0.8 -2929.4845  
3011       0.7 -3399.9800  
9774       0.8 -3701.892

## 📊 **10. Summarize Data**

Di pandas, method `DataFrame.describe()` digunakan untuk menghasilkan statistik deskriptif dari DataFrame. Metode ini sangat berguna untuk dengan cepat memahami statistik dasar dari data. Berikut adalah statistik utama yang disediakan :

* `Jumlah` : Jumlah nilai non-null dalam setiap kolom.
* `Rata-rata` : Rata-rata aritmetika dari setiap kolom.
* `Standar Deviasi` : Ukuran dispersi nilai-nilai sekitar rata-rata.
* `Minimum` : Nilai minimum dalam setiap kolom.
* `Kuartil 25 (Q1)` : Nilai di bawahnya 25% data berada.
* `Kuartil 50 (Median atau Q2)` : Nilai di bawahnya 50% data berada (yaitu nilai tengah).
* `Kuartil 75 (Q3)` : Nilai di bawahnya 75% data berada.
* `Maksimum` : Nilai maksimum dalam setiap kolom.

### 👨‍💻 Task 10

Pada DataFrame `data_superstore`, lakukan perhitungan statistik deskriptif pada kolom - kolom yang memiliki tipe data numerik

In [43]:
# Write your code below here & exec!
dt_superstore\
  .select_dtypes(include = 'number')\
  .drop(columns = ['Postal Code'])\
  .describe()

Sales     Quantity     Discount       Profit
count   9994.000000  9994.000000  9994.000000  9994.000000
mean     229.858001     3.789574     0.156203    28.656896
std      623.245101     2.225110     0.206452   234.260108
min        0.444000     1.000000     0.000000 -6599.978000
25%       17.280000     2.000000     0.000000     1.728750
50%       54.490000     3.000000     0.200000     8.666500
75%      209.940000     5.000000     0.200000    29.364000
max    22638.480000    14.000000     0.800000  8399.976000

## 🎎 **11. Aggregation**

Agregasi data merupakan serangkaian proses yang dilakukan untuk melakukan pengumpulan data dan dinyatakan dalam bentuk ringkasan sebagai contoh menghitung total yang dikelompokkan berdasarkan kategori tertentu, dsb.

Untuk melakukan agregasi gunakan sintaks berikut

<p align="center"><code><b>DataFrame.group_by(by, as_index).agg(kolom_hasil_aggregasi = (kolom_dihitung, jenis_agregasi))</b></code></p>

Jenis agregasi :
* pd.Series.sum = Menghitung total data
* pd.Series.mean = Menghitung rataan data
* pd.Series.max = Menghitung nilai maksimum data
* pd.Series.min = Menghitung nilai minimum data
* pd.Series.nunique = Menghitung jumlah data unik data
* pd.Series.count = Menghitung banyak data

### 👨‍💻 Task 11 - A

Pada DataFrame `data_superstore`, Category produk apakah yang menghasilkan total profit paling tinggi? dan berapa total profit yang didapatkan?

In [44]:
# Write your code below here & exec!
dt_superstore.groupby(['Category']).agg(total_profit=('Profit', 'sum'))

total_profit
Category                     
Furniture          18451.2728
Office Supplies   122490.8008
Technology        145454.9481

In [46]:
dt_superstore.groupby(['Category'])['Profit'].sum()

Category
Furniture           18451.2728
Office Supplies    122490.8008
Technology         145454.9481
Name: Profit, dtype: float64

Product Category dengan Profit paling tinggi adalah Kategori Technology sebesar 145454.9481

### 👨‍💻 Task 11 - B

Pada DataFrame `data_superstore`, lakukan perhitungan rataan Profit yang dikelompokkan berdasarkan Segment dan Category barang yang terjual

In [50]:
# Write your code below here & exec!
avg_profit_per_category = dt_superstore.groupby(['Segment', 'Category']).agg(Rataan_profit=('Profit', 'mean'))

display(avg_profit_per_category)

Rataan_profit
Segment     Category                      
Consumer    Furniture             6.281293
            Office Supplies      18.014174
            Technology           74.445646
Corporate   Furniture            11.741201
            Office Supplies      22.102923
            Technology           79.723823
Home Office Furniture            10.705465
            Office Supplies      24.034439
            Technology           89.152458

In [51]:
# Write your code below here & exec!
avg_profit_per_category = dt_superstore.groupby(['Segment', 'Category'], as_index = False).agg(Rataan_profit=('Profit', 'mean'))

display(avg_profit_per_category)

Segment         Category  Rataan_profit
0     Consumer        Furniture       6.281293
1     Consumer  Office Supplies      18.014174
2     Consumer       Technology      74.445646
3    Corporate        Furniture      11.741201
4    Corporate  Office Supplies      22.102923
5    Corporate       Technology      79.723823
6  Home Office        Furniture      10.705465
7  Home Office  Office Supplies      24.034439
8  Home Office       Technology      89.152458

**Written By :** <br>
<a href="https://www.linkedin.com/in/izza-putri-kamila-a812b6226/"><img alt="Linked In Link" src="https://img.shields.io/badge/-Izza%20P.%20Kamila-0072b1?style=for-the-badge&logo=None&logoColor=white" align="left"/></a>

<br><br>**for :**

<a href="https://dqlab.id/"><img src="https://dqlab.id/files/dqlab/cache/87e30118ebba5ec7d96f6ea8c9dcc10b_x_118_X_55.png" align="left" /></a>